File: SwitchBasemapSample/SwitchBasemapSample/Controllers/BasemapsListViewController.m
// Copyright 2014 ESRI
//
// All rights reserved under the copyright laws of the United States
// and applicable international laws, treaties, and conventions.
//
// You may freely redistribute and use this sample code, with or
// without modification, provided you include the original copyright
// notice and use restrictions.
//
// See the use restrictions at http://help.arcgis.com/en/sdk/10.0/usageRestrictions.htm
//

#import "BasemapsListViewController.h"
#import "SVProgressHUD.h"
#import "AppConstants.h"

@interface BasemapsListViewController ()

@property (nonatomic, strong) NSArray *portalItems;
@property (nonatomic, strong) PortalBasemapHelper *portalBasemapHelper;
@property (nonatomic, weak) IBOutlet UIButton *loadButton;
@property (nonatomic, weak) IBOutlet UIView *footerView;
@property (nonatomic, strong) AGSCredential *credential;
@property (nonatomic, assign) BOOL thumbnailLoaded;
@property (nonatomic, strong) AGSWebMap *selectedWebMap;
@property (nonatomic, strong) NSMutableDictionary *basemapDictionary;

@end

@implementation BasemapsListViewController

//in order to keep a single instance of the view controller
//creating a sharedInstance variable and will assign it the
//view controller once created
static id sharedInstance = nil;

//once the view controller is instantiated through the storyboard
//the following method is called for initialization
- (id)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        //storing the instance created as the sharedInstance inside a dispatch_once
        //block, in order to assign only the first time
        static dispatch_once_t onceToken;
        dispatch_once(&onceToken, ^{
            sharedInstance = self;
        });
    }
    return self;
}

//the following method is called everytime an object is instantiated using the initwithcoder method
//and we are using this method to return the sharedInstance (if exists) instead of the newly instantiated object
-(id)awakeAfterUsingCoder:(NSCoder *)aDecoder {
    if (sharedInstance) {
        return sharedInstance;
    }
    return self;
}

// in iOS7 this gets called and hides the status bar so the view does not go under the top iPhone status bar
- (BOOL)prefersStatusBarHidden
{
    return YES;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    //initialize the basedictionary if nil
    if (!self.basemapDictionary) {
        self.basemapDictionary = [NSMutableDictionary dictionary];
    }
}

-(void)viewDidAppear:(BOOL)animated {
    //viewDidLoad called everytime you trigger the segue
    //so in order to avoid reloading data everytime add a condition
    if (!self.portalItems) {
        self.thumbnailLoaded = NO;
        [self loadBasemaps];
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)loadBasemaps {
    //show the progress hud
    [SVProgressHUD showWithStatus:@"Loading"];
    
    //instantiate the portalBasemapHelper if nil
    if (!self.portalBasemapHelper) {
        self.portalBasemapHelper = [[PortalBasemapHelper alloc] init];
        [self.portalBasemapHelper setDelegate:self];
    }
    //prepare for the connection to portal
    NSURL *portalUrl = [NSURL URLWithString:kPortalUrl];
    //if required create a credential
    //self.credential = [[AGSCredential alloc] initWithUser:kUserName password:kPassword];
    self.credential = nil;
    [self.portalBasemapHelper fetchWebmapsFromPortal:portalUrl withCredential:self.credential];
}

//hide footer in case no more results
- (void)hideFooter {
    [self.footerView setHidden:YES];
}

//show footer if there are more results
- (void)showFooter {
    [self.footerView setHidden:NO];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    // Return the number of rows in the section.
    return self.portalItems.count;
}

-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    NSString *reusableString = @"TableCell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:reusableString];
    
    if (!cell) {
        cell = [[UITableViewCell alloc] init];
    }
    
    AGSPortalItem *item = [self.portalItems objectAtIndex:indexPath.item];
    
    if (self.thumbnailLoaded) {
        UIImageView *imageView = (UIImageView*)[cell viewWithTag:1];
        [imageView setImage:item.thumbnail];
    }
    
    UILabel *label = (UILabel*)[cell viewWithTag:2];
    [label setText:item.title];
    UILabel *descriptionLabel = (UILabel*)[cell viewWithTag:3];
    [descriptionLabel setText:item.snippet];
    
    //disable highlighting of selected cell
    [cell setSelectionStyle:UITableViewCellSelectionStyleNone];
    return cell;
}

#pragma mark - Table view delegate methods

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    AGSPortalItem *item = (AGSPortalItem*)[self.portalItems objectAtIndex:indexPath.item];
    AGSWebMapBaseMap *basemap = [self cachedBasemapForItemId:item.itemId];
    if (basemap) {
        [self.delegate basemapPickerController:self didSelectBasemap:basemap];
    }
    else {
        self.selectedWebMap = [[AGSWebMap alloc] initWithPortalItem:item];
        [self.selectedWebMap setDelegate:self];
    }
}

#pragma mark - PortalBasemapHelperDelegate methods

-(void)portalBasemapHelper:(PortalBasemapHelper *)portalBasemapHelper didFailToLoadBasemapItemsWithError:(NSError *)error {
    [[[UIAlertView alloc] initWithTitle:@"Error" message:error.localizedDescription delegate:nil cancelButtonTitle:@"Ok" otherButtonTitles: nil] show];
    [SVProgressHUD dismiss];
}

-(void)portalBasemapHelper:(PortalBasemapHelper *)portalBasemapHelper didFinishLoadingBasemapItems:(NSArray *)itemsArray {
    //populate the portalItems array and reload data
    self.portalItems = itemsArray;
    [self.tableView reloadData];
    
    //check if there are more results and hide/show the footer view accordingly
    if ([portalBasemapHelper hasMoreResults]) {
        [self showFooter];
    }
    else {
        [self hideFooter];
    }
    
    [SVProgressHUD dismiss];
}

-(void)portalBasemapHelperDidFinishFetchingThumbnails:(PortalBasemapHelper *)portalBasemapHelper {
    self.thumbnailLoaded = YES;
    //once all the thumbnails are downloaded, reload data
    [self.tableView reloadData];
}

#pragma mark - WebMap delegate methods

-(void)webMapDidLoad:(AGSWebMap *)webMap {
    //cache the base map
    [self.basemapDictionary setValue:webMap.baseMap forKeyPath:webMap.portalItem.itemId];
    [self.delegate basemapPickerController:self didSelectBasemap:self.selectedWebMap.baseMap];
}

#pragma mark - actions

-(IBAction)cancel:(id)sender {
    [self.delegate basemapPickerControllerDidCancel:self];
}

//load more results
-(IBAction)loadMoreResults:(id)sender {
    [self.portalBasemapHelper fetchNextResults];
}

#pragma mark - local/cached basemap methods

-(AGSWebMapBaseMap*)cachedBasemapForItemId:(NSString*)itemId {
    AGSWebMapBaseMap *basemap = [self.basemapDictionary objectForKey:itemId];
    return basemap;
}

@end

File: SwitchBasemapSample/swift/SwitchBasemap/Controllers/BasemapsListViewController.swift
//
// Copyright 2014 ESRI
//
// All rights reserved under the copyright laws of the United States
// and applicable international laws, treaties, and conventions.
//
// You may freely redistribute and use this sample code, with or
// without modification, provided you include the original copyright
// notice and use restrictions.
//
// See the use restrictions at http://help.arcgis.com/en/sdk/10.0/usageRestrictions.htm
//

import Foundation
import UIKit
import ArcGIS

class BasemapsListViewController:UIViewController, UITableViewDataSource, UITableViewDelegate, PortalBasemapHelperDelegate {
    
    //shared instance of the helper class
    var portalBasemapHelper = PortalBasemapHelper.sharedInstance
    //delegate that handles selection actions
    weak var delegate:BasemapPickerDelegate?
    //footer view to provide feedback if there are more webmaps
    @IBOutlet weak var footerView: UIView!
    //to list the webmaps
    @IBOutlet weak var tableView: UITableView!
    
    
    //to hide the status bar when presenting this view controller
    override func prefersStatusBarHidden() -> Bool {
        return true
    }
    
    override func viewDidLoad() {
        //assign the delegate as self for helper class
        portalBasemapHelper.delegate = self
        //if webmaps not already present, call the method on the helper class
        if portalBasemapHelper.webmapsArray.count == 0 {
            showHUDWithStatus("Loading")
            portalBasemapHelper.fetchWebmaps(kPortalUrl, credential: nil)
        }
    }
    
    //MARK: footer view methods
    
    //to hide the footer view
    func hideFooterView() {
        self.footerView.hidden = true
    }
    //to show the footer view
    func showFooterView() {
        self.footerView.hidden = false
    }
    
    //MARK: table view data source
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        //check if there are more results and hide/show the footer view accordingly
        if self.portalBasemapHelper.hasMoreResults() {
            self.showFooterView()
        }
        else {
            self.hideFooterView()
        }
        
       return portalBasemapHelper.webmapsArray.count;
    }
    
    //using the title of the portal item as the label, snippet for description and thumbnail as image view
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("TableCell")!
        
        let webmap = portalBasemapHelper.webmapsArray[indexPath.row]
        
        let titleLabel = cell.viewWithTag(2) as! UILabel
        titleLabel.text = webmap.title
        
        let descriptionLabel = cell.viewWithTag(3) as! UILabel
        descriptionLabel.text = webmap.snippet
        
        let thumbnailImageView = cell.viewWithTag(1) as! UIImageView
        //Use the thumbnail image if there else use the placeholder image
        if webmap.thumbnail != nil {
            thumbnailImageView.image = webmap.thumbnail
        }
        else {
            thumbnailImageView.image = UIImage(named: "Placeholder")
        }
        
        return cell
    }
    
    //MARK: table view delegates
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let portalItem = self.portalBasemapHelper.webmapsArray[indexPath.row]
        //check if there is a cached version of the basemap
        //if yes, then notify the delegate
        //else initiate the process for downloading and caching the basemap
        if let basemap = self.portalBasemapHelper.cachedBasemap(forItemId: portalItem.itemId) {
            self.delegate?.basemapPickerController(self, didSelectBasemap: basemap)
        }
        else {
            self.portalBasemapHelper.cacheBasemap(forPortalItem:portalItem)
        }
    }
    
    //MARK: portal basemap helper delegates
    
    func portalBasemapHelper(portalBasemapHelper:PortalBasemapHelper, didFailWithError error: NSError) {
        dismissHUD()
        //display any error from the helper class in the alert view
        UIAlertView(title: "Error", message: error.localizedDescription, delegate: nil, cancelButtonTitle: "Ok").show()
    }
    
    func portalBasemapHelper(portalBasemapHelper:PortalBasemapHelper, didFinishLoadingPortalItems portalItems:[AGSPortalItem]) {
        self.tableView.reloadData()
        
        dismissHUD()
    }
    
    func portalBasemapHelper(portalBasemapHelper: PortalBasemapHelper, didFetchThumbnail thumbnail: UIImage, forPortalItem portalItem: AGSPortalItem) {
        //find the index of the portalItem in the array
        
        if let index = self.portalBasemapHelper.webmapsArray.indexOf(portalItem) {
            //get the array of visible cells
            let visibleCells = self.tableView.visibleCells
            //get the cell corresponding to the index
            if let cell = self.tableView.cellForRowAtIndexPath(NSIndexPath(forRow: index, inSection: 0)) {
                //check if the cell is visible
                let isVisible = visibleCells.contains(cell)
                //and if visible then update the image
                if isVisible {
                    self.tableView.reloadRowsAtIndexPaths([NSIndexPath(forRow: index, inSection: 0)], withRowAnimation:.None)
                }
            }
        }
    }
    
    func portalBasemapHelper(portalBasemapHelper: PortalBasemapHelper, didFinishCachingBasemap basemap: AGSWebMapBaseMap, forWebMap webMap: AGSWebMap)  {
        //notify the delegate about the selected basemap
        self.delegate?.basemapPickerController(self, didSelectBasemap: basemap)
    }
    
    //MARK: actions
    
    //notify the delegate to hide the controller if tapped on cancel
    @IBAction func cancelAction(sender: UIBarButtonItem) {
        self.delegate?.basemapPickerControllerDidCancel(self)
    }
    
    //fetch next results if tapped on more
    @IBAction func loadMoreResults(sender:AnyObject) {
        showHUDWithStatus("Loading")
        self.portalBasemapHelper.fetchNextResults()
    }
}